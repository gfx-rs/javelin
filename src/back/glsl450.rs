use crate::{
    back::glsl_common::{write_type, Error, StatementBuilder},
    Binding, BuiltIn, FastHashMap, Module, TypeInner,
};
use std::{
    fmt::{self},
    io::Write,
};

pub fn write(module: &Module, out: &mut impl Write) -> Result<(), Error> {
    writeln!(out, "#version 450 core")?;

    let mut counter = 0;
    let mut names = FastHashMap::default();

    let mut namer = |name: Option<&String>| {
        if let Some(name) = name {
            names.insert(name.clone(), ());
            name.clone()
        } else {
            counter += 1;
            while names.get(&format!("_{}", counter)).is_some() {
                counter += 1;
            }
            format!("_{}", counter)
        }
    };

    let mut structs = FastHashMap::default();

    // Do a first pass to collect names
    for (handle, ty) in module.types.iter() {
        match ty.inner {
            TypeInner::Struct { ref members } => {
                let name = namer(ty.name.as_ref());
                let mut idxs = Vec::new();

                for member in members {
                    idxs.push(namer(member.name.as_ref()))
                }

                structs.insert(handle, (name, idxs));
            }
            _ => continue,
        }
    }

    // Do a second pass to build the structs
    // TODO: glsl is order dependent so we need to build structs in order
    for (handle, ty) in module.types.iter() {
        match ty.inner {
            TypeInner::Struct { ref members } => {
                let (name, idxs) = structs.get(&handle).unwrap();

                writeln!(out, "struct {} {{", name)?;
                for (member, name) in members.iter().zip(idxs.iter()) {
                    writeln!(
                        out,
                        "   {} {}",
                        write_type(member.ty, &module.types, &structs)?,
                        name
                    )?;
                }
                writeln!(out, "}};")?;
            }
            _ => continue,
        }
    }

    let mut globals_lookup = FastHashMap::default();

    for (handle, global) in module.global_variables.iter() {
        if let Some(Binding::BuiltIn(built_in)) = global.binding {
            let semantic = match built_in {
                BuiltIn::Position => "gl_position",
                BuiltIn::GlobalInvocationId => "gl_GlobalInvocationID",
                BuiltIn::BaseInstance => "gl_BaseInstance",
                BuiltIn::BaseVertex => "gl_BaseVertex",
                BuiltIn::ClipDistance => "gl_ClipDistance",
                BuiltIn::InstanceIndex => "gl_InstanceIndex",
                BuiltIn::VertexIndex => "gl_VertexIndex",
                BuiltIn::PointSize => "gl_PointSize",
                BuiltIn::FragCoord => "gl_FragCoord",
                BuiltIn::FrontFacing => "gl_FrontFacing",
                BuiltIn::SampleIndex => "gl_SampleID",
                BuiltIn::FragDepth => "gl_FragDepth",
                BuiltIn::LocalInvocationId => "gl_LocalInvocationID",
                BuiltIn::LocalInvocationIndex => "gl_LocalInvocationIndex",
                BuiltIn::WorkGroupId => "gl_WorkGroupID",
            };

            globals_lookup.insert(handle, String::from(semantic));
            continue;
        }

        if let Some(ref binding) = global.binding {
            write!(out, "layout({}) ", binding.writer())?;
        }

        let name = namer(global.name.as_ref());

        writeln!(
            out,
            "{}{} {};",
            global.class.write_glsl(),
            write_type(global.ty, &module.types, &structs)?,
            name
        )?;

        globals_lookup.insert(handle, name);
    }

    let mut functions = FastHashMap::default();

    // Do a first pass to collect names
    for (handle, func) in module.functions.iter() {
        functions.insert(handle, namer(func.name.as_ref()));
    }

    let mut types = module.types.clone();
    let mut typifier = crate::proc::Typifier::new();

    // TODO: glsl is order dependent so we need to build functions in order
    for (handle, func) in module.functions.iter() {
        let name = functions.get(&handle).unwrap();

        writeln!(
            out,
            "{} {}({}) {{",
            func.return_type
                .map_or(Ok(String::from("void")), |ty| write_type(
                    ty, &types, &structs
                ))?,
            name,
            func.parameter_types
                .iter()
                .map(|ty| write_type(*ty, &types, &structs))
                .collect::<Result<Vec<_>, _>>()?
                .join(","),
        )?;

        let mut builder = StatementBuilder {
            functions: &functions,
            globals: &globals_lookup,
            locals_lookup: &func
                .local_variables
                .iter()
                .map(|(handle, local)| (handle, namer(local.name.as_ref())))
                .collect(),
            structs: &structs,
            args: &func
                .parameter_types
                .iter()
                .enumerate()
                .map(|(pos, _)| (pos as u32, namer(None)))
                .collect(),
            expressions: &func.expressions,
            types: &mut types,
            locals: &func.local_variables,
            typifier: &mut typifier,
        };

        for sta in func.body.iter() {
            writeln!(out, "{}", sta.write_glsl(module, &mut builder)?)?;
        }

        writeln!(out, "}}")?;
    }

    Ok(())
}

struct BindingWriter<'a> {
    inner: &'a Binding,
}

impl Binding {
    fn writer<'a>(&'a self) -> BindingWriter<'a> {
        BindingWriter { inner: self }
    }
}

impl<'a> fmt::Display for BindingWriter<'a> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self.inner {
            Binding::BuiltIn(_) => write!(f, ""), // Ignore because they are variables with a predefined name
            Binding::Location(location) => write!(f, "location={}", location),
            Binding::Descriptor { set, binding } => write!(f, "set={},binding={}", set, binding),
        }
    }
}
