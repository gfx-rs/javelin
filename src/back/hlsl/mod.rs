mod keywords;
mod writer;

use std::fmt::Error as FmtError;
use thiserror::Error;

pub use writer::Writer;

#[derive(Debug, Copy, Clone, PartialEq, PartialOrd)]
pub struct ShaderModel(u16);

impl Default for ShaderModel {
    fn default() -> Self {
        ShaderModel(50)
    }
}

#[derive(Error, Debug)]
pub enum Error {
    #[error(transparent)]
    IoError(#[from] FmtError),
    #[error("BuiltIn {0:?} is not supported")]
    UnsupportedShaderModel(ShaderModel),
    #[error("A scalar with an unsupported width was requested: {0:?} {1:?}")]
    UnsupportedScalar(crate::ScalarKind, crate::Bytes),
    #[error("BuiltIn {0:?} is not supported")]
    UnsupportedBuiltIn(crate::BuiltIn),
    #[error("{0}")]
    Unimplemented(String), // TODO: Error used only during development
}

pub fn write_string(
    module: &crate::Module,
    info: &crate::valid::ModuleInfo,
    shader_model: ShaderModel,
) -> Result<String, Error> {
    let mut w = Writer::new(String::new(), shader_model);
    w.write(module, info)?;
    let output = w.finish();
    Ok(output)
}
