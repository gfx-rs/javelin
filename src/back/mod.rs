//! Functions which export shader modules into binary and text formats.

#[cfg(feature = "glsl-out")]
pub mod glsl;
#[cfg(feature = "msl-out")]
pub mod msl;
#[cfg(feature = "spv-out")]
pub mod spv;

impl crate::Expression {
    /// Returns the ref count, upon reaching which this expression
    /// should be considered for baking.
    #[allow(dead_code)]
    fn bake_ref_count(&self) -> usize {
        match *self {
            // accesses are never cached, only loads are
            crate::Expression::Access { .. } | crate::Expression::AccessIndex { .. } => !0,
            // image operations look better when isolated
            crate::Expression::ImageSample { .. } | crate::Expression::ImageLoad { .. } => 1,
            // cache expressions that are referenced multiple times
            _ => 2,
        }
    }
}
