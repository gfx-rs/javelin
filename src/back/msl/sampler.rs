#[derive(Clone, Debug, PartialEq)]
pub enum Coord {
    Normalized,
    Pixel,
}

impl Default for Coord {
    fn default() -> Self {
        Self::Normalized
    }
}

#[derive(Clone, Debug, PartialEq)]
pub enum Address {
    Repeat,
    MirroredRepeat,
    ClampToEdge,
    ClampToZero,
    ClampToBorder,
}

impl Default for Address {
    fn default() -> Self {
        Self::ClampToEdge
    }
}

#[derive(Clone, Debug, PartialEq)]
pub enum BorderColor {
    TransparentBlack,
    OpaqueBlack,
    OpaqueWhite,
}

impl Default for BorderColor {
    fn default() -> Self {
        Self::TransparentBlack
    }
}

#[derive(Clone, Debug, PartialEq)]
pub enum Filter {
    Nearest,
    Linear,
}

impl Default for Filter {
    fn default() -> Self {
        Self::Nearest
    }
}

#[derive(Clone, Debug, PartialEq)]
pub enum CompareFunc {
    Never,
    Less,
    LessEqual,
    Greater,
    GreaterEqual,
    Equal,
    NotEqual,
    Always,
}

impl Default for CompareFunc {
    fn default() -> Self {
        Self::Never
    }
}

#[derive(Clone, Debug, Default, PartialEq)]
pub struct InlineSampler {
    pub coord: Coord,
    pub address: [Address; 3],
    pub border_color: BorderColor,
    pub mag_filter: Filter,
    pub min_filter: Filter,
    pub mip_filter: Option<Filter>,
    pub compare_func: CompareFunc,
}
