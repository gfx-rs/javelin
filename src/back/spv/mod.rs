mod layout;
pub mod parser;

use spirv::*;

struct Module {
    physical_layout: PhysicalLayout,
    logical_layout: LogicalLayout,
}

impl Module {
    fn new(header: &crate::Header) -> Self {
        Module {
            physical_layout: PhysicalLayout::new(header, 0x0u32),
            logical_layout: LogicalLayout::new(),
        }
    }
}

struct PhysicalLayout {
    magic_number: Word,
    version: Word,
    generator: Word,
    bound: Word,
    instruction_schema: Word,
}

impl PhysicalLayout {
    fn new(header: &crate::Header, instruction_schema: Word) -> Self {
        let version: Word = (0x0u32 << 24)
            | ((header.version.0 as u32) << 16)
            | ((header.version.1 as u32) << 8)
            | header.version.2 as u32;

        PhysicalLayout {
            magic_number: MAGIC_NUMBER,
            version,
            generator: header.generator,
            bound: 0,
            instruction_schema,
        }
    }

    fn in_words(&self) -> Vec<Word> {
        let mut words: Vec<Word> = vec![];

        words.push(self.magic_number);
        words.push(self.version);
        words.push(self.generator);

        match self.bound {
            0 => panic!("Bound cannot be 0"),
            _ => words.push(self.bound),
        }

        words.push(self.instruction_schema);
        words
    }
}

pub(crate) struct LogicalLayout {
    capabilities: Vec<Word>,
    extensions: Vec<Word>,
    ext_inst_imports: Vec<Word>,
    memory_model: Vec<Word>,
    entry_points: Vec<Word>,
    execution_modes: Vec<Word>,
    debugs: Vec<Word>,
    annotations: Vec<Word>,
    type_declarations: Vec<Word>,
    constants: Vec<Word>,
    global_variables: Vec<Word>,
    function_declarations: Vec<Word>,
    function_definitions: Vec<Word>,
}

impl LogicalLayout {
    fn new() -> Self {
        LogicalLayout {
            capabilities: vec![],
            extensions: vec![],
            ext_inst_imports: vec![],
            memory_model: vec![],
            entry_points: vec![],
            execution_modes: vec![],
            debugs: vec![],
            annotations: vec![],
            type_declarations: vec![],
            constants: vec![],
            global_variables: vec![],
            function_declarations: vec![],
            function_definitions: vec![],
        }
    }

    fn in_words(&self) -> Vec<u32> {
        let mut words: Vec<Word> = vec![];

        for capability in self.capabilities.iter() {
            words.push(*capability);
        }

        for extension in self.extensions.iter() {
            words.push(*extension);
        }

        for ext_inst_import in self.ext_inst_imports.iter() {
            words.push(*ext_inst_import);
        }

        for memory_model in self.memory_model.iter() {
            words.push(*memory_model);
        }

        for entry_point in self.entry_points.iter() {
            words.push(*entry_point);
        }

        for execution_mode in self.execution_modes.iter() {
            words.push(*execution_mode);
        }

        for debug in self.debugs.iter() {
            words.push(*debug);
        }

        for annotation in self.annotations.iter() {
            words.push(*annotation);
        }

        for type_declaration in self.type_declarations.iter() {
            words.push(*type_declaration);
        }

        for constants in self.constants.iter() {
            words.push(*constants);
        }

        for global_variables in self.global_variables.iter() {
            words.push(*global_variables);
        }

        for function_declaration in self.function_declarations.iter() {
            words.push(*function_declaration);
        }

        for function_definition in self.function_definitions.iter() {
            words.push(*function_definition);
        }

        words
    }
}

struct Instruction {
    op: Op,
    wc: u32,
    type_id: Option<Word>,
    result_id: Option<Word>,
    operands: Vec<Word>,
}

impl Instruction {
    fn new(op: Op) -> Self {
        Instruction {
            op,
            wc: 1, // Always start at 1 for the first word (OP + WC),
            type_id: None,
            result_id: None,
            operands: vec![],
        }
    }

    fn set_type(&mut self, id: Word) {
        assert!(self.type_id.is_none(), "Type can only be set once");
        self.type_id = Some(id);
        self.wc += 1;
    }

    fn set_result(&mut self, id: Word) {
        assert!(self.result_id.is_none(), "Result can only be set once");
        self.result_id = Some(id);
        self.wc += 1;
    }

    fn add_operand(&mut self, operand: Word) {
        self.operands.push(operand);
        self.wc += 1;
    }

    fn add_operands(&mut self, operands: Vec<Word>) {
        for operand in operands {
            self.add_operand(operand);
        }
    }

    fn to_words(&self) -> Vec<Word> {
        let mut words = Vec::with_capacity(self.wc as usize);

        let wc_op = (self.wc << 16 | self.op as u32) as u32;

        words.push(wc_op);

        if self.type_id.is_some() {
            words.push(self.type_id.unwrap());
        }

        if self.result_id.is_some() {
            words.push(self.result_id.unwrap());
        }

        for operand in self.operands.iter() {
            words.push(*operand);
        }

        words
    }
}
