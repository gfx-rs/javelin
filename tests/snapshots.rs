use serde::Deserialize;

#[derive(Hash, PartialEq, Eq, Deserialize)]
enum Stage {
    Vertex,
    Fragment,
    Compute,
}

#[derive(Hash, PartialEq, Eq, Deserialize)]
struct BindSource {
    stage: Stage,
    group: u32,
    binding: u32,
}

#[derive(Deserialize)]
struct BindTarget {
    #[serde(default)]
    buffer: Option<u8>,
    #[serde(default)]
    texture: Option<u8>,
    #[serde(default)]
    sampler: Option<u8>,
    #[serde(default)]
    mutable: bool,
}

#[derive(Default, Deserialize)]
struct Parameters {
    #[serde(default)]
    spv_flow_dump_prefix: String,
    #[cfg_attr(not(feature = "spv-out"), allow(dead_code))]
    spv_capabilities: naga::FastHashSet<spirv::Capability>,
    #[cfg_attr(not(feature = "msl-out"), allow(dead_code))]
    mtl_bindings: naga::FastHashMap<BindSource, BindTarget>,
}

#[cfg(feature = "msl-out")]
fn check_output_msl(module: &naga::Module, name: &str, params: &Parameters) {
    use naga::back::msl;
    let mut binding_map = msl::BindingMap::default();
    for (key, value) in params.mtl_bindings.iter() {
        binding_map.insert(
            msl::BindSource {
                stage: match key.stage {
                    Stage::Vertex => naga::ShaderStage::Vertex,
                    Stage::Fragment => naga::ShaderStage::Fragment,
                    Stage::Compute => naga::ShaderStage::Compute,
                },
                group: key.group,
                binding: key.binding,
            },
            msl::BindTarget {
                buffer: value.buffer,
                texture: value.texture,
                sampler: value.sampler,
                mutable: value.mutable,
            },
        );
    }
    let options = msl::Options {
        lang_version: (1, 0),
        spirv_cross_compatibility: false,
        binding_map,
    };
    let (msl, _) = msl::write_string(&module, &options).unwrap();
    insta::assert_snapshot!(format!("{}{}", name, ".msl"), msl);
}

#[cfg(feature = "spv-out")]
fn check_output_spv(module: &naga::Module, name: &str, params: &Parameters) {
    use naga::back::spv;
    use rspirv::binary::Disassemble;
    let spv = spv::write_vec(
        &module,
        spv::WriterFlags::NONE,
        params.spv_capabilities.clone(),
    )
    .unwrap();
    let dis = rspirv::dr::load_words(spv)
        .expect("Produced invalid SPIR-V")
        .disassemble();
    insta::assert_snapshot!(format!("{}{}", name, ".spvasm"), dis);
}

#[cfg(feature = "wgsl-in")]
fn convert_wgsl(name: &str) {
    let params =
        match std::fs::read_to_string(format!("tests/snapshots/in/{}{}", name, ".param.ron")) {
            Ok(string) => ron::de::from_str(&string).expect("Couldn't find param file"),
            Err(_) => Parameters::default(),
        };

    let module = naga::front::wgsl::parse_str(
        &std::fs::read_to_string(format!("tests/snapshots/in/{}{}", name, ".wgsl"))
            .expect("Couldn't find wgsl file"),
    )
    .unwrap();
    naga::proc::Validator::new().validate(&module).unwrap();

    #[cfg(feature = "msl-out")]
    check_output_msl(&module, name, &params);
    #[cfg(feature = "spv-out")]
    check_output_spv(&module, name, &params);
}

#[cfg(feature = "wgsl-in")]
#[test]
fn converts_wgsl_quad() {
    convert_wgsl("quad");
}
