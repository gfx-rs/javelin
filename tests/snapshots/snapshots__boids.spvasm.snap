---
source: tests/snapshots.rs
expression: dis
---
; SPIR-V
; Version: 1.0
; Generator: rspirv
; Bound: 298
OpCapability Shader
%1 = OpExtInstImport "GLSL.std.450"
OpMemoryModel Logical GLSL450
OpEntryPoint GLCompute %29 "main" %36
OpExecutionMode %29 LocalSize 64 1 1
OpDecorate %36 BuiltIn GlobalInvocationId
OpMemberDecorate %45 0 Offset 0
OpMemberDecorate %45 1 Offset 8
OpDecorate %47 ArrayStride 16
OpDecorate %49 BufferBlock
OpMemberDecorate %49 0 Offset 0
OpDecorate %50 NonWritable
OpDecorate %50 DescriptorSet 0
OpDecorate %50 Binding 1
OpDecorate %121 Block
OpMemberDecorate %121 0 Offset 0
OpMemberDecorate %121 1 Offset 4
OpMemberDecorate %121 2 Offset 8
OpMemberDecorate %121 3 Offset 12
OpMemberDecorate %121 4 Offset 16
OpMemberDecorate %121 5 Offset 20
OpMemberDecorate %121 6 Offset 24
OpDecorate %122 DescriptorSet 0
OpDecorate %122 Binding 0
OpDecorate %274 DescriptorSet 0
OpDecorate %274 Binding 2
%3 = OpTypeInt 32 1
%2 = OpConstant  %3  1500
%5 = OpTypeFloat 32
%4 = OpConstant  %5  0.0
%6 = OpConstant  %3  0
%8 = OpTypeInt 32 0
%7 = OpConstant  %8  0
%9 = OpConstant  %3  1
%10 = OpConstant  %8  1
%11 = OpConstant  %5  1.0
%12 = OpConstant  %5  0.1
%13 = OpConstant  %5  -1.0
%15 = OpTypeVector %5 2
%16 = OpTypePointer Function %15
%22 = OpTypePointer Function %3
%27 = OpTypePointer Function %8
%28 = OpTypeVoid
%30 = OpTypeFunction %28
%32 = OpTypeBool
%35 = OpTypeVector %8 3
%37 = OpTypePointer Input %35
%36 = OpVariable  %37  Input
%38 = OpTypePointer Input %8
%39 = OpConstant  %3  0
%45 = OpTypeStruct %15 %15
%47 = OpTypeRuntimeArray %45
%49 = OpTypeStruct %47
%51 = OpTypePointer Uniform %49
%50 = OpVariable  %51  Uniform
%52 = OpTypePointer Uniform %47
%53 = OpConstant  %3  0
%55 = OpTypePointer Input %8
%56 = OpConstant  %3  0
%58 = OpTypePointer Uniform %45
%59 = OpTypePointer Uniform %15
%60 = OpConstant  %3  0
%65 = OpTypePointer Uniform %47
%66 = OpConstant  %3  0
%68 = OpTypePointer Input %8
%69 = OpConstant  %3  0
%71 = OpTypePointer Uniform %45
%72 = OpTypePointer Uniform %15
%73 = OpConstant  %3  1
%90 = OpTypePointer Input %8
%91 = OpConstant  %3  0
%99 = OpTypePointer Uniform %47
%100 = OpConstant  %3  0
%102 = OpTypePointer Uniform %45
%103 = OpTypePointer Uniform %15
%104 = OpConstant  %3  0
%109 = OpTypePointer Uniform %47
%110 = OpConstant  %3  0
%112 = OpTypePointer Uniform %45
%113 = OpTypePointer Uniform %15
%114 = OpConstant  %3  1
%121 = OpTypeStruct %5 %5 %5 %5 %5 %5 %5
%123 = OpTypePointer Uniform %121
%122 = OpVariable  %123  Uniform
%124 = OpTypePointer Uniform %5
%125 = OpConstant  %3  1
%140 = OpTypePointer Uniform %5
%141 = OpConstant  %3  2
%156 = OpTypePointer Uniform %5
%157 = OpConstant  %3  3
%198 = OpTypePointer Uniform %5
%199 = OpConstant  %3  4
%204 = OpTypePointer Uniform %5
%205 = OpConstant  %3  5
%210 = OpTypePointer Uniform %5
%211 = OpConstant  %3  6
%224 = OpTypePointer Uniform %5
%225 = OpConstant  %3  0
%229 = OpTypePointer Function %5
%230 = OpConstant  %3  0
%236 = OpTypePointer Function %5
%237 = OpConstant  %3  0
%240 = OpTypePointer Function %5
%241 = OpConstant  %3  0
%247 = OpTypePointer Function %5
%248 = OpConstant  %3  0
%251 = OpTypePointer Function %5
%252 = OpConstant  %3  1
%258 = OpTypePointer Function %5
%259 = OpConstant  %3  1
%262 = OpTypePointer Function %5
%263 = OpConstant  %3  1
%269 = OpTypePointer Function %5
%270 = OpConstant  %3  1
%274 = OpVariable  %51  Uniform
%275 = OpTypePointer Uniform %47
%276 = OpConstant  %3  0
%278 = OpTypePointer Input %8
%279 = OpConstant  %3  0
%281 = OpTypePointer Uniform %45
%282 = OpTypePointer Uniform %15
%283 = OpConstant  %3  0
%288 = OpTypePointer Uniform %47
%289 = OpConstant  %3  0
%291 = OpTypePointer Input %8
%292 = OpConstant  %3  0
%294 = OpTypePointer Uniform %45
%295 = OpTypePointer Uniform %15
%296 = OpConstant  %3  1
%29 = OpFunction  %28  None %30
%31 = OpLabel
%26 = OpVariable  %27  Function %7
%23 = OpVariable  %22  Function %6
%19 = OpVariable  %16  Function
%14 = OpVariable  %16  Function
%24 = OpVariable  %16  Function
%20 = OpVariable  %16  Function
%17 = OpVariable  %16  Function
%25 = OpVariable  %16  Function
%21 = OpVariable  %22  Function %6
%18 = OpVariable  %16  Function
%34 = OpAccessChain  %38  %36 %39
%40 = OpLoad  %8  %34
%33 = OpUGreaterThanEqual  %32  %40 %2
OpSelectionMerge %41 None
OpBranchConditional %33 %42 %43
%42 = OpLabel
OpReturn
%43 = OpLabel
OpBranch %41
%41 = OpLabel
%48 = OpAccessChain  %52  %50 %53
%54 = OpAccessChain  %55  %36 %56
%57 = OpLoad  %8  %54
%46 = OpAccessChain  %58  %48 %57
%44 = OpAccessChain  %59  %46 %60
%61 = OpLoad  %15  %44
OpStore %14 %61
%64 = OpAccessChain  %65  %50 %66
%67 = OpAccessChain  %68  %36 %69
%70 = OpLoad  %8  %67
%63 = OpAccessChain  %71  %64 %70
%62 = OpAccessChain  %72  %63 %73
%74 = OpLoad  %15  %62
OpStore %17 %74
%75 = OpCompositeConstruct  %15  %4 %4
OpStore %18 %75
%76 = OpCompositeConstruct  %15  %4 %4
OpStore %19 %76
%77 = OpCompositeConstruct  %15  %4 %4
OpStore %20 %77
OpBranch %78
%78 = OpLabel
OpLoopMerge %79 %81 None
OpBranch %80
%80 = OpLabel
%83 = OpLoad  %8  %26
%82 = OpUGreaterThanEqual  %32  %83 %2
OpSelectionMerge %84 None
OpBranchConditional %82 %85 %86
%85 = OpLabel
OpBranch %79
%86 = OpLabel
OpBranch %84
%84 = OpLabel
%88 = OpLoad  %8  %26
%89 = OpAccessChain  %90  %36 %91
%92 = OpLoad  %8  %89
%87 = OpIEqual  %32  %88 %92
OpSelectionMerge %93 None
OpBranchConditional %87 %94 %95
%94 = OpLabel
OpBranch %81
%95 = OpLabel
OpBranch %93
%93 = OpLabel
%98 = OpAccessChain  %99  %50 %100
%101 = OpLoad  %8  %26
%97 = OpAccessChain  %102  %98 %101
%96 = OpAccessChain  %103  %97 %104
%105 = OpLoad  %15  %96
OpStore %24 %105
%108 = OpAccessChain  %109  %50 %110
%111 = OpLoad  %8  %26
%107 = OpAccessChain  %112  %108 %111
%106 = OpAccessChain  %113  %107 %114
%115 = OpLoad  %15  %106
OpStore %25 %115
%117 = OpLoad  %15  %24
%118 = OpLoad  %15  %14
%119 = OpExtInst  %5  %1 Distance %117 %118
%120 = OpAccessChain  %124  %122 %125
%126 = OpLoad  %5  %120
%116 = OpFOrdLessThan  %32  %119 %126
OpSelectionMerge %127 None
OpBranchConditional %116 %128 %129
%128 = OpLabel
%131 = OpLoad  %15  %18
%132 = OpLoad  %15  %24
%130 = OpFAdd  %15  %131 %132
OpStore %18 %130
%134 = OpLoad  %3  %21
%133 = OpIAdd  %3  %134 %9
OpStore %21 %133
OpBranch %127
%129 = OpLabel
OpBranch %127
%127 = OpLabel
%136 = OpLoad  %15  %24
%137 = OpLoad  %15  %14
%138 = OpExtInst  %5  %1 Distance %136 %137
%139 = OpAccessChain  %140  %122 %141
%142 = OpLoad  %5  %139
%135 = OpFOrdLessThan  %32  %138 %142
OpSelectionMerge %143 None
OpBranchConditional %135 %144 %145
%144 = OpLabel
%147 = OpLoad  %15  %20
%149 = OpLoad  %15  %24
%150 = OpLoad  %15  %14
%148 = OpFSub  %15  %149 %150
%146 = OpFSub  %15  %147 %148
OpStore %20 %146
OpBranch %143
%145 = OpLabel
OpBranch %143
%143 = OpLabel
%152 = OpLoad  %15  %24
%153 = OpLoad  %15  %14
%154 = OpExtInst  %5  %1 Distance %152 %153
%155 = OpAccessChain  %156  %122 %157
%158 = OpLoad  %5  %155
%151 = OpFOrdLessThan  %32  %154 %158
OpSelectionMerge %159 None
OpBranchConditional %151 %160 %161
%160 = OpLabel
%163 = OpLoad  %15  %19
%164 = OpLoad  %15  %25
%162 = OpFAdd  %15  %163 %164
OpStore %19 %162
%166 = OpLoad  %3  %23
%165 = OpIAdd  %3  %166 %9
OpStore %23 %165
OpBranch %159
%161 = OpLabel
OpBranch %159
%159 = OpLabel
OpBranch %81
%81 = OpLabel
%168 = OpLoad  %8  %26
%167 = OpIAdd  %8  %168 %10
OpStore %26 %167
OpBranch %78
%79 = OpLabel
%170 = OpLoad  %3  %21
%169 = OpSGreaterThan  %32  %170 %6
OpSelectionMerge %171 None
OpBranchConditional %169 %172 %173
%172 = OpLabel
%176 = OpLoad  %15  %18
%178 = OpLoad  %3  %21
%179 = OpConvertSToF  %5  %178
%177 = OpFDiv  %5  %11 %179
%175 = OpVectorTimesScalar  %15  %176 %177
%180 = OpLoad  %15  %14
%174 = OpFSub  %15  %175 %180
OpStore %18 %174
OpBranch %171
%173 = OpLabel
OpBranch %171
%171 = OpLabel
%182 = OpLoad  %3  %23
%181 = OpSGreaterThan  %32  %182 %6
OpSelectionMerge %183 None
OpBranchConditional %181 %184 %185
%184 = OpLabel
%187 = OpLoad  %15  %19
%189 = OpLoad  %3  %23
%190 = OpConvertSToF  %5  %189
%188 = OpFDiv  %5  %11 %190
%186 = OpVectorTimesScalar  %15  %187 %188
OpStore %19 %186
OpBranch %183
%185 = OpLabel
OpBranch %183
%183 = OpLabel
%194 = OpLoad  %15  %17
%196 = OpLoad  %15  %18
%197 = OpAccessChain  %198  %122 %199
%200 = OpLoad  %5  %197
%195 = OpVectorTimesScalar  %15  %196 %200
%193 = OpFAdd  %15  %194 %195
%202 = OpLoad  %15  %20
%203 = OpAccessChain  %204  %122 %205
%206 = OpLoad  %5  %203
%201 = OpVectorTimesScalar  %15  %202 %206
%192 = OpFAdd  %15  %193 %201
%208 = OpLoad  %15  %19
%209 = OpAccessChain  %210  %122 %211
%212 = OpLoad  %5  %209
%207 = OpVectorTimesScalar  %15  %208 %212
%191 = OpFAdd  %15  %192 %207
OpStore %17 %191
%214 = OpLoad  %15  %17
%215 = OpExtInst  %15  %1 Normalize %214
%216 = OpLoad  %15  %17
%217 = OpExtInst  %5  %1 Length %216
%218 = OpExtInst  %5  %1 FClamp %217 %4 %12
%213 = OpVectorTimesScalar  %15  %215 %218
OpStore %17 %213
%220 = OpLoad  %15  %14
%222 = OpLoad  %15  %17
%223 = OpAccessChain  %224  %122 %225
%226 = OpLoad  %5  %223
%221 = OpVectorTimesScalar  %15  %222 %226
%219 = OpFAdd  %15  %220 %221
OpStore %14 %219
%228 = OpAccessChain  %229  %14 %230
%231 = OpLoad  %5  %228
%227 = OpFOrdLessThan  %32  %231 %13
OpSelectionMerge %232 None
OpBranchConditional %227 %233 %234
%233 = OpLabel
%235 = OpAccessChain  %236  %14 %237
OpStore %235 %11
OpBranch %232
%234 = OpLabel
OpBranch %232
%232 = OpLabel
%239 = OpAccessChain  %240  %14 %241
%242 = OpLoad  %5  %239
%238 = OpFOrdGreaterThan  %32  %242 %11
OpSelectionMerge %243 None
OpBranchConditional %238 %244 %245
%244 = OpLabel
%246 = OpAccessChain  %247  %14 %248
OpStore %246 %13
OpBranch %243
%245 = OpLabel
OpBranch %243
%243 = OpLabel
%250 = OpAccessChain  %251  %14 %252
%253 = OpLoad  %5  %250
%249 = OpFOrdLessThan  %32  %253 %13
OpSelectionMerge %254 None
OpBranchConditional %249 %255 %256
%255 = OpLabel
%257 = OpAccessChain  %258  %14 %259
OpStore %257 %11
OpBranch %254
%256 = OpLabel
OpBranch %254
%254 = OpLabel
%261 = OpAccessChain  %262  %14 %263
%264 = OpLoad  %5  %261
%260 = OpFOrdGreaterThan  %32  %264 %11
OpSelectionMerge %265 None
OpBranchConditional %260 %266 %267
%266 = OpLabel
%268 = OpAccessChain  %269  %14 %270
OpStore %268 %13
OpBranch %265
%267 = OpLabel
OpBranch %265
%265 = OpLabel
%273 = OpAccessChain  %275  %274 %276
%277 = OpAccessChain  %278  %36 %279
%280 = OpLoad  %8  %277
%272 = OpAccessChain  %281  %273 %280
%271 = OpAccessChain  %282  %272 %283
%284 = OpLoad  %15  %14
OpStore %271 %284
%287 = OpAccessChain  %288  %274 %289
%290 = OpAccessChain  %291  %36 %292
%293 = OpLoad  %8  %290
%286 = OpAccessChain  %294  %287 %293
%285 = OpAccessChain  %295  %286 %296
%297 = OpLoad  %15  %17
OpStore %285 %297
OpReturn
OpFunctionEnd
